File: src/test/java/com/tms/csp/dnnf/Projection.java
package com.tms.csp.dnnf;

import com.tms.csp.ast.*;
import com.tms.csp.data.CspSample;
import com.tms.csp.fm.dnnf.products.Cube;
import com.tms.csp.fm.dnnf.vars.VarGrp;
import com.tms.csp.util.CspBaseTest2;
import com.tms.csp.util.varSets.VarSet;
import org.junit.Test;

import java.util.Set;

import static org.junit.Assert.assertEquals;

public class Projection extends CspBaseTest2 {


    @Test
    public void tiny() throws Exception {
        Exp n = loadDnnfTinyNoDc().getSmooth();

        assertEquals(3, n.getSatCount());
        assertEquals(3, n.getForEachSatCount());

        n.printCubes();
        System.err.println();


        n = n.project("a", "b").getSmooth();


        assertEquals(2, n.getVars().size());
        assertEquals(2, n.getSatCount());


        n.printCubes();

    }

    /*
    add(CspSample.Trim, 8, "L4", "V6", "Hybrid", "Base", "LE", "SE");
     */
    @Test
    public void trimWithOutVars() throws Exception {
        Csp csp = loadCsp(CspSample.Trim, false);

        Space space = csp.getSpace();

        Exp n1 = csp.toDnnf().getSmooth();
        VarSet outVars = space.getVars(Prefix.ENG);
        Exp n2 = n1.project(outVars);

        long cubeCount = n2.getCubes().size();
        assertEquals(3, cubeCount);
        for (Cube cube : n2.getCubes()) {
            System.err.println(cube.getTrueVars());
        }


    }

    //satCount FACTORY  [5669]    - 5788
    //satCount INV      [6564]
    @Test
    public void testEfc() throws Exception {
//        Exp n = loadTinyDnnf(dnnfEfc);
        Exp n = loadTinyDnnf(efcOriginalDnnf);

        Space space = n.getSpace();


        Set<Var> vars = space.getVars(VarGrp.CORE);

        n = n.project(vars);
        Exp smooth = n.getSmooth();

        long satCount = smooth.getSatCount();
        System.err.println("satCount[" + satCount + "]");


        assertEquals(5788, satCount);


    }

}

File: src/test/java/com/tms/csp/dnnf/Projection.kt
package com.tms.csp.dnnf

import com.tms.csp.ast.Csp
import com.tms.csp.ast.toCubes
import com.tms.csp.data.CspSample
import com.tms.csp.fm.dnnf.vars.VarGrp
import org.junit.Assert.assertEquals
import org.junit.Test

class Projection {


    @Test
    fun tiny() {
        var n = Csp.compileDnnf(CspSample.TinyNoDc).smooth

        assertEquals(3L, n.satCountLong)
        assertEquals(3, n.forEachSatCount.toLong())

        n.printCubes()
        System.err.println()


        n = n.project("a", "b").smooth


        assertEquals(2, n.vars.size)
        assertEquals(2L, n.satCountLong)


        n.printCubes()

    }


    @Test
    fun trimWithOutVars() {

        val expectedCubes3 = """
            L4 Base
            L4 LE
            V6 LE
            L4 XLE
            V6 XLE
            L4 SE
            V6 SE
            Hybrid Hyb
        """.trimIndent()

        val grades = "Base Hyb LE SE XLE"
        val engines = "Hybrid L4 V6"
        //        Csp csp = loadCsp(CspSample.TrimNoDc, false);
        val csp = Csp.parse(CspSample.TrimNoDc)

        val space = csp.space

        val n1 = csp.toDnnf().smooth
        val engVars = space.mkVarSet(engines)
        val grdVars = space.mkVarSet(grades)
        val grdEngVars = engVars.plus(grdVars)

        System.err.println("engVars[$engVars]")
        System.err.println("grdVars[$grdVars]")
        System.err.println("grdEngVars[$grdEngVars]")
        System.err.println()

        val n2 = n1.project(engVars)

        val cubeCount = n2.cubes.size.toLong()
        assertEquals(3, cubeCount)
        for (cube in n2.cubes) {
            System.err.println(cube.trueVars)
        }
        System.err.println()


        val n3 = n1.project(grdEngVars)


        val products3 = expectedCubes3.toCubes(space).map { it.trueVars }

        val cubes3 = n3.cubes.map { it.trueVars }
        assertEquals(8, cubes3.size)
        assertEquals(products3, cubes3)

    }

    @Test
    fun testEfc() {
        var n = Csp.compileDnnf(CspSample.EfcOriginal)


        val space = n.space

        val vars = space.getVars(VarGrp.CORE)

        n = n.project(vars)
        val smooth = n.smooth

        val satCount = smooth.satCount.toLong()


        assertEquals(5788, satCount)


    }

}